motion/dp_controller/src/dp_controller_ros.cpp

```diff
@@ -0,0 +1,254 @@
+          position, orientation, velocity, nu_d, eta_d_pos, eta_d_ori);
+
+      Eigen::Vector6d DOF = Eigen::Vector6d::Zero();
+      int i = 0;
+      for (int j : dp_server.goal_.DOF) {
+        tau(i) *= j;
+        DOF(i) = dp_server.goal_.DOF[i];
+        i++;
+      }
+      tf::wrenchEigenToMsg(tau, tau_msg);
+
+      m_wrench_pub.publish(tau_msg);
+      m_referencepoint_pub.publish(dp_server.goal_.x_ref);
+
+      //----------------- DEBUG -------------------
+
+      std_msgs::Float64MultiArray P_debug_msg;
+      std_msgs::Float64MultiArray I_debug_msg;
+      std_msgs::Float64MultiArray D_debug_msg;
+      for (int i = 0; i < 6; i++) {
+        P_debug_msg.data.push_back(m_controller.P_debug(i));
+        I_debug_msg.data.push_back(m_controller.I_debug(i));
+        D_debug_msg.data.push_back(m_controller.D_debug(i));
+      }
+
+      m_dp_P_debug_pub.publish(P_debug_msg);
+      m_dp_I_debug_pub.publish(I_debug_msg);
+      m_dp_D_debug_pub.publish(D_debug_msg);
+
+      //-------------------- END DEBUG -------------------
+    }
+
+    Eigen::Vector3d orientation_euler = QuaterniondToEuler(orientation);
+    dp_server.pose << position, orientation_euler;
+
+    // Makes sure DP always ends by sending 0 thrust
+    if (was_active && !dp_server.run_controller) {
+      Eigen::VectorXd tau_zero = Eigen::VectorXd::Zero(6, 1);
+      tf::wrenchEigenToMsg(tau_zero, tau_msg);
+      m_wrench_pub.publish(tau_msg);
+      was_active = false;
+    }
+
+    ros::spinOnce();
+    rate.sleep();
+  }
+}
+
+void Controller::odometryCallback(const nav_msgs::Odometry &msg) {
+  // Convert to eigen for computation
+  tf::pointMsgToEigen(msg.pose.pose.position, position);
+  tf::quaternionMsgToEigen(msg.pose.pose.orientation, orientation);
+  tf::twistMsgToEigen(msg.twist.twist, velocity);
+}
+
+int sgn(double x) {
+  if (x < 0)
+    return -1;
+  return 1;
+}
+
+void Controller::desiredPointCallback(const nav_msgs::Odometry &desired_msg) {
+
+  // Convert to eigen for computation
+  tf::pointMsgToEigen(desired_msg.pose.pose.position, eta_d_pos);
+  tf::quaternionMsgToEigen(desired_msg.pose.pose.orientation, eta_d_ori);
+  tf::twistMsgToEigen(desired_msg.twist.twist, nu_d);
+
+  // // --------------------------- DEBUG ----------------------
+  // std_msgs::Float32 debug_msg;
+  // Eigen::Vector3d Debug_vec = QuaterniondToEuler(eta_d_ori);
+  // debug_msg.data = Debug_vec(0) * 180 / M_PI;
+  // m_reference_return_DEBUG_pub.publish(debug_msg);
+
+  // Eigen::Quaterniond q_tilde = eta_d_ori.conjugate() * orientation;
+  // std_msgs::Float32 debug2_msg;
+  // debug2_msg.data = QuaterniondToEuler(q_tilde)(0) * 180 / M_PI;
+  // m_reference_return_DEBUG2_pub.publish(debug2_msg);
+
+  // std_msgs::Float32 q_tilde_print;
+  // q_tilde_print.data = sgn(q_tilde.w()) * q_tilde.x();
+  // m_reference_return_q_tilde_print_pub.publish(q_tilde_print);
+  // // ----------------------------- END DEBUG
+  // ---------------------------------
```

Comment:
Remove comments if unused, or make it toggleable if it will be used in the future